staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,334 @@
+    Write-ConsoleLog "gpu cert url is set to $GpuDriverCertURL"
+  
+    # Set some default values
+    $Driver = @{
+        # Hard coding the FwLink for resources.json
+        InstallExe   = "install.exe"
+        RebootNeeded = $false
+    }
+    $Index = @{
+        OS = 0 # Windows
+    }
+  
+    try {
+        $Compute = Get-VmData
+        $vmSize = $Compute.vmSize
+    }
+    catch {
+        $ErrorMsg = "Failed to query the SKU information."
+        Write-ConsoleLog $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_SKU_INFO_NOT_FOUND -ErrorMessage $ErrorMsg
+    }
+  
+    # Not an AzureStack scenario
+    if ($vmSize -ne $null) {
+        if ( ($Compute.vmSize -Match "_NC") -or
+          ($Compute.vmSize -Match "_ND") ) {
+            # Override if GRID driver is desired on ND or NC VMs
+            $Index.Driver = 0 # CUDA
+        }
+        elseif ( $Compute.vmSize -Match "_NV" ) {
+            # NV or Grid on ND or Grid on NC
+            $Index.Driver = 1 # GRID
+            $Driver.RebootNeeded = $true
+        }
+        else {
+            $ErrorMsg = "VM type $($Compute.vmSize) is not an N-series VM. Not attempting to deploy extension! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_VM_SIZE_NOT_SUPPORTED -ErrorMessage $ErrorMsg
+        }
+    }
+  
+    # Get the certificate url
+    $Driver.CertificateUrl = $GpuDriverCertURL
+    $Driver.Url = $GpuDriverURL
+  
+    return $Driver
+}
+  
+function Get-VmData {
+    [OutputType([hashtable])]
+  
+    # Retry to overcome failure in retrieving VM metadata
+    $Loop = $true
+    $RetryCount = 0
+    $RetryCountMax = 10
+    do {
+        try {
+            $Compute = Invoke-RestMethod -Headers @{"Metadata" = "true" } -URI http://169.254.169.254/metadata/instance/compute?api-version=2017-08-01 -Method get
+            $Loop = $false
+        }
+        catch {
+            if ($RetryCount -gt $RetryCountMax) {
+                $Loop = $false
+                $ErrorMsg = "Failed to retrieve VM metadata after $RetryCountMax attempts. Exiting! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+                Write-ConsoleLog $ErrorMsg
+                Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_SKU_INFO_NOT_FOUND -ErrorMessage $ErrorMsg
+            }
+            else {
+                Start-Sleep -Seconds ($RetryCount * 2 + 1)
+                $RetryCount++
+            }
+        }
+    } while ($Loop)
+  
+    return $Compute
+}
+  
+function Get-DriverFile {
+    param(
+        [Parameter(Mandatory = $True)]
+        [ValidateNotNullOrEmpty()]
+        [string] $source,
+    
+        [Parameter(Mandatory = $True)]
+        [ValidateNotNullOrEmpty()]
+        [string] $dest
+    )
+    
+    function GetUsingWebClient {
```

Comment:
We use DownloadFileOverHttp so we can avoid download when the file is cached in the vhd.